### PR TITLE
Added Metrics rate limits

### PR DESCRIPTION
**Pull request by jamescrosswell**
Resolves https://github.com/getsentry/sentry-dotnet/issues/3248

### CODE REVIEW

File: test/Sentry.Tests/Internals/Http/HttpTransportTests.cs [hunk 1]
```diff
@@ -276,13 +276,16 @@ public async Task SendEnvelopeAsync_ResponseNotOkNoMessage_LogsError()
         ).Should().BeTrue();
     }
 
-    [Fact]
-    public async Task SendEnvelopeAsync_ItemRateLimit_DropsItem()
+    [Theory]
+    [InlineData("2700:metric_bucket:organization:quota_exceeded:custom", true)] // Default namespace... we back off
+    [InlineData("2700:metric_bucket:organization:quota_exceeded", true)] // No namespace... we back off
```
Comment (cleptric) [inline test/Sentry.Tests/Internals/Http/HttpTransportTests.cs:281]: Let's add

```c#
[InlineData("2700:metric_bucket:organization:", true)] // Empty namespace... we back off
```

File: src/Sentry/Internal/Http/RateLimitCategory.cs [hunk 2]
```diff
@@ -17,13 +17,18 @@ public bool Matches(EnvelopeItem item)
...
             return false;
         }
 
-        return string.Equals(Name, type, StringComparison.OrdinalIgnoreCase);
+        // The envelope item type for metrics is `statsd`, but the client report category is `metric_bucket`
+        var reportCategory = EnvelopeItem.TypeValueMetric == type
+            ? "metric_bucket"
+            : type;
+
+        return string.Equals(Name, reportCategory, StringComparison.OrdinalIgnoreCase);
```
Comment (bitsandfoxes) [inline src/Sentry/Internal/Http/RateLimitCategory.cs:31]: Can we be a bit more opinionated here? We're now allocating a new string, comparing multiple times in different ways just to special case the `metric_bucket -> statsd` conversion.
And that happens for every category.

File: src/Sentry/Internal/Http/RateLimit.cs [hunk 3]
```diff
@@ -21,10 +26,19 @@ public static RateLimit Parse(string rateLimitEncoded)
...
         var categories = components[1].Split(';').Select(c => new RateLimitCategory(c)).ToArray();
+        string[]? namespaces = null;
+        foreach (var category in categories)
+        {
+            if (!string.Equals(category.Name, "metric_bucket", StringComparison.OrdinalIgnoreCase))
+            {
+                continue;
+            }
+
+            namespaces = components.Length > 4 ? components[4].Split(';') : null;
```
Comment (bitsandfoxes) [inline src/Sentry/Internal/Http/RateLimit.cs:37]:
```suggestion
            // Response header looking like this: X-Sentry-Rate-Limits: 2700:metric_bucket:organization:quota_exceeded:custom
            namespaces = components.Length > 4 ? components[4].Split(';') : null;
```

File: src/Sentry/Internal/Http/RateLimitCategory.cs (hunk 2 above)
Comment (jamescrosswell) [reply src/Sentry/Internal/Http/RateLimitCategory.cs]: I've replaced with a switch that removes the string allocation.

File: src/Sentry/Internal/Http/RateLimitCategory.cs [hunk 4]
```diff
@@ -17,13 +17,20 @@ public bool Matches(EnvelopeItem item)
             return true;
         }
 
-        var type = item.TryGetType();
+        var type = item.TryGetType()?.ToLower();
```
Comment (bitsandfoxes) [inline src/Sentry/Internal/Http/RateLimitCategory.cs:20]:
```suggestion
        var type = item.TryGetType();
```
I think this is now redundant.

**Review by bitsandfoxes — APPROVED** [review]
Thanks!

**Pull request merged**